Python/importdl.c

```diff
@@ -205,6 +210,142 @@ _Py_ext_module_loader_info_init_from_spec(
+    case _Py_ext_module_loader_result_EXCEPTION:  /* fall through */
+    case _Py_ext_module_loader_result_ERR_MISSING:  /* fall through */
+    case _Py_ext_module_loader_result_ERR_UNREPORTED_EXC:  /* fall through */
+    case _Py_ext_module_loader_result_ERR_NONASCII_NOT_MULTIPHASE:  /* fall through */
+    case _Py_ext_module_loader_result_ERR_NOT_MODULE:  /* fall through */
+    case _Py_ext_module_loader_result_ERR_MISSING_DEF:
+        break;
+    default:
+        /* We added a new error kind but forgot to add it to this switch. */
+        assert(0);
+    }
+}
+
+void
+_Py_ext_module_loader_result_apply_error(
+                            struct _Py_ext_module_loader_result *res,
+                            const char *name)
+{
+    assert(!PyErr_Occurred());
+    struct _Py_ext_module_loader_result_error *err = res->err;
+    assert(err != NULL && err == &res->_err);
+
+#ifndef NDEBUG
+    switch (err->kind) {
+    case _Py_ext_module_loader_result_EXCEPTION:  /* fall through */
+    case _Py_ext_module_loader_result_ERR_UNREPORTED_EXC:
+        assert(err->exc != NULL);
+        break;
+    case _Py_ext_module_loader_result_ERR_MISSING:  /* fall through */
+    case _Py_ext_module_loader_result_ERR_UNINITIALIZED:  /* fall through */
+    case _Py_ext_module_loader_result_ERR_NONASCII_NOT_MULTIPHASE:  /* fall through */
+    case _Py_ext_module_loader_result_ERR_NOT_MODULE:  /* fall through */
+    case _Py_ext_module_loader_result_ERR_MISSING_DEF:
+        assert(err->exc == NULL);
+        break;
+    default:
+        /* We added a new error kind but forgot to add it to this switch. */
+        assert(0);
+    }
+#endif
+
+    const char *msg = NULL;
+    switch (err->kind) {
+    case _Py_ext_module_loader_result_EXCEPTION:
+        break;
+    case _Py_ext_module_loader_result_ERR_MISSING:
+        msg = "initialization of %s failed without raising an exception";
+        break;
+    case _Py_ext_module_loader_result_ERR_UNREPORTED_EXC:
+        msg = "initialization of %s raised unreported exception";
+        break;
+    case _Py_ext_module_loader_result_ERR_UNINITIALIZED:
+        msg = "init function of %s returned uninitialized object";
+        break;
+    case _Py_ext_module_loader_result_ERR_NONASCII_NOT_MULTIPHASE:
+        msg = "initialization of %s did not return PyModuleDef";
+        break;
+    case _Py_ext_module_loader_result_ERR_NOT_MODULE:
+        msg = "initialization of %s did not return an extension module";
+        break;
+    case _Py_ext_module_loader_result_ERR_MISSING_DEF:
+        msg = "initialization of %s did not return a valid extension module";
+        break;
+    default:
+        /* We added a new error kind but forgot to add it to this switch. */
+        assert(0);
+        PyErr_Format(PyExc_SystemError,
+                     "loading %s failed due to init function", name);
+        return;
+    }
+
+    if (err->exc != NULL) {
+        PyErr_SetRaisedException(err->exc);
```

Comment:
Consider warning the reader about that function's quirk:
```suggestion
        PyErr_SetRaisedException(err->exc);
        err->exc = NULL; /* SetRaisedException stole our reference */
```
Maybe it would be better to merge `_Py_ext_module_loader_result_apply_error` and `_Py_ext_module_loader_result_clear`? They're effectively always used together (with `apply` skipped on success).

Reply:
done